libs/utils/src/crashsafe.rs

```diff
@@ -123,15 +123,21 @@ pub async fn fsync_async_opt(
     Ok(())
 }
 
-/// Like postgres' durable_rename, renames file issuing fsyncs do make it
-/// durable. After return, file and rename are guaranteed to be persisted.
+/// Like postgres' durable_rename, renames a file and issues fsyncs to make it durable. After
+/// returning, both the file and rename are guaranteed to be persisted. Both paths must be on the
+/// same file system.
 ///
-/// Unlike postgres, it only does fsyncs to 1) file to be renamed to make
-/// contents durable; 2) its directory entry to make rename durable 3) again to
-/// already renamed file, which is not required by standards but postgres does
-/// it, let's stick to that. Postgres additionally fsyncs newpath *before*
-/// rename if it exists to ensure that at least one of the files survives, but
-/// current callers don't need that.
+/// Unlike postgres, it only fsyncs 1) the file to make contents durable, and 2) the directory to
+/// make the rename durable. This sequence ensures the target file will never be incomplete.
+///
+/// Postgres also:
+///
+/// * Fsyncs the target file, if it exists, before the rename, to ensure either the new or existing
+///   file survives a crash. Current callers don't need this as it should already be fsynced if
+///   durability is needed.
+///
+/// * Fsyncs the file after the rename. This can be required with certain OSes or file systems (e.g.
+///   NFS), but not on Linux with common file systems like ext4 (which we currently use).
```

Comment:
Indeed. But still dunno. I can't find quickly anything enough affirmative, and chatgpt says

"""
There are indeed cases where certain filesystems or configurations could necessitate the second `fsync` for durability in `durable_rename`. Here are some specific examples:

1. **ext4 with `data=writeback` or `data=ordered` modes**:
   - The `ext4` filesystem has different journaling modes that influence how metadata and data are handled.
   - In `data=writeback` mode, metadata updates are journaled, but data writes are not ordered with respect to metadata updates. This means that metadata might be persisted while data could still be in cache.
   - Even in `data=ordered` mode, where data writes are ordered before metadata commits, a rename might not be fully durable across crashes without an extra `fsync` on the file itself after the rename.
   - The final `fsync` ensures that the file's metadata, as modified by the rename, is fully committed to disk.

2. **XFS**:
   - `XFS` is known for its high performance and scalability, but its metadata journaling is sometimes not fully crash-proof when only a directory `fsync` is performed.
   - While `XFS` is generally reliable, some versions have demonstrated inconsistencies where, after a crash, the rename operation wasn't entirely durable with only a directory `fsync`.
   - Performing a final `fsync` on the file itself ensures that all operations related to the file’s new state are completed.

3. **NFS (Network File System)**:
   - With NFS, especially in asynchronous or unreliable network configurations, directory and file `fsync`s don't always behave as expected due to client-server caching behaviors.
   - Some NFS setups require explicit control over individual file and metadata updates. Relying solely on a directory `fsync` may leave room for inconsistency.
   - A post-rename `fsync` on the file mitigates this risk, helping to ensure that the rename survives network disruptions or server crashes.

4. **Btrfs**:
   - `Btrfs` has its own set of complexities due to its copy-on-write (COW) nature. While it has good durability features, certain Btrfs configurations or kernel versions have shown issues where renames aren’t fully consistent across crashes if not followed by a second `fsync` on the file.
   - The copy-on-write behavior sometimes causes metadata to be inconsistently updated if only the directory `fsync` is used.

5. **ZFS**:
   - In some cases, `ZFS` on Linux might display similar behaviors, particularly when running on top of hardware or virtualized environments that don’t have full disk-level durability guarantees.
   - ZFS generally handles metadata durably, but ensuring the rename operation is fully committed with a final `fsync` on the file itself is a good precaution in case of unexpected power loss or system crashes.

In summary, while many modern filesystems aim to make renames durable with minimal `fsync` operations, differences in journaling, metadata handling, and caching strategies make the post-rename `fsync` an added safeguard, especially in environments where durability is critical.
"""

So I wouldn't want to go there.

Reply:
I think ChatGPT is getting confused here -- it seems to not understand that the file was fsynced before the rename.

I'm almost certain this is sufficient with ext4 and XFS, and I'd be surprised if it weren't with Btrfs and ZFS as it's such a common idiom. Sure, there are configurations where it may not hold, but those are configurations that explicitly trade off durability for performance (like noop'ing fsyncs), and not something we'd consider using.

Let me get us some more evidence to build confidence that this is safe:

* Write a test case demonstrating (via kernel diagnostics) that buffers are fully flushed to disk after each step.
* Run the test case on ext4, XFS, Btrfs, and ZFS.
* Find examples of what other databases do.
* Get opinions of other Postgres people.
* Run the benchmark on a representative Linux box, to gauge what the real benefit would be.

I'm waiting for SREs to provision a Linux box for me.

Reply:
I tested this on ext4, using eBPF probes to look at block device operations. It confirmed that the last fsync doesn't do anything.

I set up a loopback block device with ext4, and ran the following Rust binary and eBPF probes:

<details>
<summary>syncrename.rs</summary>

```rs
fn main() -> Result<(), Box<dyn std::error::Error>> {
    // Parse arguments.
    let args: Vec<String> = std::env::args().collect();
    if args.len() != 2 {
        eprintln!("Usage: {} <dir>", args[0]);
        std::process::exit(1);
    }
    let dirpath = std::path::PathBuf::from(&args[1]);
    let sourcepath = dirpath.join("source");
    let targetpath = dirpath.join("target");

    // Create the source file.
    std::fs::write(&sourcepath, vec![0; 1_000_000])?;

    // Fsync the source file.
    std::fs::File::open(&sourcepath)?.sync_all()?;

    // Rename the file.
    std::fs::rename(&sourcepath, &targetpath)?;

    // Fsync the dir.
    std::fs::File::open(&dirpath)?.sync_all()?;

    // Fsync the target file.
    std::fs::File::open(&targetpath)?.sync_all()?;

    Ok(())
}
```
</details>

<details>
<summary>trace.ebpf</summary>

```
tracepoint:block:block_rq_issue {
    printf("block req pid=%d process=%s dev=%d:%d rwbs=%s bytes=%d\n", pid, comm, args->dev >> 20, args-> dev & 0xfffff, args->rwbs, args->bytes);
}

tracepoint:syscalls:sys_enter_fsync /comm == "syncrename"/ {
    printf("fsync call pid=%d process=%s fd=%d\n", pid, comm, args->fd);
}

tracepoint:syscalls:sys_exit_fsync /comm == "syncrename"/ {
    printf("fsync done pid=%d process=%s return=%d\n", pid, comm, args->ret);
}

tracepoint:syscalls:sys_enter_write /comm == "syncrename"/ {
    printf("write call pid=%d process=%s fd=%d\n", pid, comm, args->fd);
}

tracepoint:syscalls:sys_exit_write /comm == "syncrename"/ {
    printf("write done pid=%d process=%s return=%d\n", pid, comm, args->ret);
}

tracepoint:syscalls:sys_enter_rename /comm == "syncrename"/ {
    printf("rename call pid=%d process=%s old=%s new=%s\n", pid, comm, str(args->oldname), str(args->newname));
}
```
</details>

Here's the annotated eBPF output:

```
# Write 1MB to file `source`.
write call pid=128589 process=syncrename fd=3
write done pid=128589 process=syncrename return=1000000

# Fsync file `source`. Block device flags are:
# WS = write sync
# WSM = write sync metadata
# FF = flush
fsync call pid=128589 process=syncrename fd=3
block req pid=128589 process=syncrename dev=7:0 rwbs=WS bytes=479232
block req pid=128589 process=syncrename dev=7:0 rwbs=WS bytes=524288
block req pid=120725 process=jbd2/loop0-8 dev=7:0 rwbs=WSM bytes=4096
block req pid=120725 process=jbd2/loop0-8 dev=7:0 rwbs=WSM bytes=4096
block req pid=120725 process=jbd2/loop0-8 dev=7:0 rwbs=WSM bytes=4096
block req pid=120725 process=jbd2/loop0-8 dev=7:0 rwbs=WSM bytes=4096
block req pid=120725 process=jbd2/loop0-8 dev=7:0 rwbs=WSM bytes=4096
block req pid=120725 process=jbd2/loop0-8 dev=7:0 rwbs=WSM bytes=4096
block req pid=377 process=kworker/24:1H dev=7:0 rwbs=FF bytes=0
block req pid=372 process=kworker/25:1H dev=7:0 rwbs=WSM bytes=4096
block req pid=372 process=kworker/25:1H dev=7:0 rwbs=FF bytes=0
fsync done pid=128589 process=syncrename return=0

# Rename the file.
rename call pid=128589 process=syncrename old=disk/source new=disk/target

# Fsync directory.
fsync call pid=128589 process=syncrename fd=3
block req pid=120725 process=jbd2/loop0-8 dev=7:0 rwbs=WSM bytes=4096
block req pid=120725 process=jbd2/loop0-8 dev=7:0 rwbs=WSM bytes=4096
block req pid=120725 process=jbd2/loop0-8 dev=7:0 rwbs=WSM bytes=4096
block req pid=120725 process=jbd2/loop0-8 dev=7:0 rwbs=WSM bytes=4096
block req pid=120725 process=jbd2/loop0-8 dev=7:0 rwbs=WSM bytes=4096
block req pid=120725 process=jbd2/loop0-8 dev=7:0 rwbs=WSM bytes=4096
block req pid=120725 process=jbd2/loop0-8 dev=7:0 rwbs=WSM bytes=4096
block req pid=377 process=kworker/24:1H dev=7:0 rwbs=FF bytes=0
block req pid=372 process=kworker/25:1H dev=7:0 rwbs=WSM bytes=4096
block req pid=372 process=kworker/25:1H dev=7:0 rwbs=FF bytes=0
fsync done pid=128589 process=syncrename return=0

# Fsync target file.
fsync call pid=128589 process=syncrename fd=3
block req pid=533 process=kworker/6:1H dev=7:0 rwbs=FF bytes=0
fsync done pid=128589 process=syncrename return=0
```

As you can see, the final fsync only sends an `FF` flush operation to the block device, it doesn't write anything. The block device was already flushed after the directory fsync above, so this flush is a noop.

I'll rerun this experiment with XFS and Btrfs after lunch, but I'm pretty sure they'll show the same result.

Reply:
## XFS

XFS behavior is the same as ext4 (with some additional reads due to a cold cache):

```
# Write 1MB of data.
block req pid=131533 process=syncrename dev=7:0 rwbs=RM bytes=4096
block req pid=131533 process=syncrename dev=7:0 rwbs=RM bytes=4096
write call pid=131533 process=syncrename fd=3
write done pid=131533 process=syncrename return=1000000

# fsync `source`.
fsync call pid=131533 process=syncrename fd=3
block req pid=131533 process=syncrename dev=7:0 rwbs=RM bytes=512
block req pid=131533 process=syncrename dev=7:0 rwbs=RM bytes=4096
block req pid=131533 process=syncrename dev=7:0 rwbs=RM bytes=4096
block req pid=131533 process=syncrename dev=7:0 rwbs=RM bytes=4096
block req pid=131533 process=syncrename dev=7:0 rwbs=WS bytes=1003520
block req pid=359 process=kworker/26:1H dev=7:0 rwbs=FF bytes=0
block req pid=366 process=kworker/10:1H dev=7:0 rwbs=WSM bytes=2560
block req pid=366 process=kworker/10:1H dev=7:0 rwbs=FF bytes=0
fsync done pid=131533 process=syncrename return=0

# Rename.
rename call pid=131533 process=syncrename old=disk/source new=disk/target

# Fsync dir.
fsync call pid=131533 process=syncrename fd=3
block req pid=359 process=kworker/26:1H dev=7:0 rwbs=FF bytes=0
block req pid=366 process=kworker/10:1H dev=7:0 rwbs=WSM bytes=1536
block req pid=366 process=kworker/10:1H dev=7:0 rwbs=FF bytes=0
fsync done pid=131533 process=syncrename return=0

# Fsync `target`.
fsync call pid=131533 process=syncrename fd=3
block req pid=359 process=kworker/26:1H dev=7:0 rwbs=FF bytes=0
fsync done pid=131533 process=syncrename return=0
```

## ZFS

ZFS also does the same as ext4 (except it doesn't send a noop flush to the block device):

```
# Write 1 MB.
write call pid=134460 process=syncrename fd=3
write done pid=134460 process=syncrename return=1000000

# Fsync `source`.
fsync call pid=134460 process=syncrename fd=3
block req pid=134461 process=z_wr_iss_h dev=7:0 rwbs=W bytes=4096
block req pid=377 process=kworker/24:1H dev=7:0 rwbs=FF bytes=0
fsync done pid=134460 process=syncrename return=0

# Rename.
rename call pid=134460 process=syncrename old=/loop/disk/source new=/loop/disk/target

# Fsync dir.
fsync call pid=134460 process=syncrename fd=3
block req pid=134050 process=z_wr_iss_h dev=7:0 rwbs=W bytes=4096
block req pid=377 process=kworker/24:1H dev=7:0 rwbs=FF bytes=0
fsync done pid=134460 process=syncrename return=0

# Fsync `target`.
fsync call pid=134460 process=syncrename fd=3
fsync done pid=134460 process=syncrename return=0
```

# Btrfs

With Btrfs the `target` fsync does result in additional writes. An additional, second fsync of `target` doesn't result in any block device operations:

```
# Write 1MB of data.
write call pid=132275 process=syncrename fd=3
write done pid=132275 process=syncrename return=1000000

# fsync `source`.
fsync call pid=132275 process=syncrename fd=3
block req pid=132275 process=syncrename dev=7:0 rwbs=WS bytes=479232
block req pid=132275 process=syncrename dev=7:0 rwbs=WS bytes=524288
block req pid=132275 process=syncrename dev=7:0 rwbs=WSM bytes=16384
block req pid=132275 process=syncrename dev=7:0 rwbs=WSM bytes=16384
block req pid=132275 process=syncrename dev=7:0 rwbs=WSM bytes=16384
block req pid=132275 process=syncrename dev=7:0 rwbs=WSM bytes=16384
block req pid=533 process=kworker/6:1H dev=7:0 rwbs=FF bytes=0
block req pid=132275 process=syncrename dev=7:0 rwbs=WSM bytes=4096
block req pid=418 process=kworker/5:1H dev=7:0 rwbs=FF bytes=0
fsync done pid=132275 process=syncrename return=0

# Rename.
rename call pid=132275 process=syncrename old=disk/source new=disk/target

# Fsync dir.
fsync call pid=132275 process=syncrename fd=3
block req pid=132275 process=syncrename dev=7:0 rwbs=WSM bytes=16384
block req pid=132275 process=syncrename dev=7:0 rwbs=WSM bytes=16384
block req pid=132275 process=syncrename dev=7:0 rwbs=WSM bytes=16384
block req pid=132275 process=syncrename dev=7:0 rwbs=WSM bytes=16384
block req pid=533 process=kworker/6:1H dev=7:0 rwbs=FF bytes=0
block req pid=132275 process=syncrename dev=7:0 rwbs=WSM bytes=4096
block req pid=418 process=kworker/5:1H dev=7:0 rwbs=FF bytes=0
fsync done pid=132275 process=syncrename return=0

# Fsync `target`.
fsync call pid=132275 process=syncrename fd=3
block req pid=132275 process=syncrename dev=7:0 rwbs=WSM bytes=16384
block req pid=132275 process=syncrename dev=7:0 rwbs=WSM bytes=16384
block req pid=132275 process=syncrename dev=7:0 rwbs=WSM bytes=16384
block req pid=132275 process=syncrename dev=7:0 rwbs=WSM bytes=16384
block req pid=533 process=kworker/6:1H dev=7:0 rwbs=FF bytes=0
block req pid=132275 process=syncrename dev=7:0 rwbs=WSM bytes=4096
block req pid=418 process=kworker/5:1H dev=7:0 rwbs=FF bytes=0
fsync done pid=132275 process=syncrename return=0

# Fsync `target` again.
fsync call pid=132275 process=syncrename fd=3
fsync done pid=132275 process=syncrename return=0
```

It isn't clear what these writes are or if they're necessary for durability. In an [email thread](https://www.spinics.net/lists/linux-btrfs/msg77330.html), Chris Mason (the Btrfs author) discourages fsyncing the directory entirely (for performance), but says that a directory fsync will make all the directory contents durable as well. However, that email thread was started specifically to discuss data loss with fsync(file) + rename + fsync(dir) -- maybe they hit [this bug](https://patchwork.kernel.org/project/linux-btrfs/patch/1455294046-8768-1-git-send-email-fdmanana@kernel.org/) which lost files after renames.

So I guess let's not run Neon on Btrfs and we should be fine.

Reply:
Thanks! I'll add a comment about Btrfs and the above findings.

Reply:
Not much -- I am good with fsync'ing file and fsync'ing the parent only.  This stuff is frustratingly poorly defined: we'll probably always be able to find some filesystem that does something strange (especially network filesystems), but all I care about is how we run on ext4 and XFS.